backend/core/src/main/java/org/sonarsource/sonarlint/core/tracking/matching/MatchingSession.java

```diff
@@ -0,0 +1,119 @@
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.tracking.matching;
+
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.sonarsource.sonarlint.core.analysis.RawIssue;
+import org.sonarsource.sonarlint.core.commons.KnownFinding;
+import org.sonarsource.sonarlint.core.commons.RuleType;
+import org.sonarsource.sonarlint.core.tracking.IntroductionDateProvider;
+import org.sonarsource.sonarlint.core.tracking.IssueMapper;
+import org.sonarsource.sonarlint.core.tracking.KnownFindings;
+import org.sonarsource.sonarlint.core.tracking.TextRangeUtils;
+import org.sonarsource.sonarlint.core.tracking.TrackedIssue;
+
+public class MatchingSession {
+  private final Map<Path, List<KnownFinding>> remainingUnmatchedIssuesPerFile;
+  private final Map<Path, List<KnownFinding>> remainingUnmatchedSecurityHotspotsPerFile;
+  private final IntroductionDateProvider introductionDateProvider;
+  private final ConcurrentHashMap<Path, List<TrackedIssue>> issuesPerFile = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<Path, List<TrackedIssue>> securityHotspotsPerFile = new ConcurrentHashMap<>();
+  private final Set<Path> relativePathsInvolved = new HashSet<>();
+
+  public MatchingSession(KnownFindings previousFindings, IntroductionDateProvider introductionDateProvider) {
+    this.remainingUnmatchedIssuesPerFile = previousFindings.getIssuesPerFile().entrySet().stream().map(entry -> Map.entry(entry.getKey(), new ArrayList<>(entry.getValue())))
+      .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    this.remainingUnmatchedSecurityHotspotsPerFile = previousFindings.getSecurityHotspotsPerFile().entrySet().stream()
+      .map(entry -> Map.entry(entry.getKey(), new ArrayList<>(entry.getValue()))).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    this.introductionDateProvider = introductionDateProvider;
+  }
+
+  public TrackedIssue matchWithKnownFinding(Path relativePath, RawIssue rawIssue) {
+    if (RuleType.SECURITY_HOTSPOT.equals(rawIssue.getRuleType())) {
```

Comment:
Maybe introduce `isSecurityHotspot()` method as in `TrackedIssue`

Reply:
Good suggestion :+1: